graphql_compiler/schema_generation/schema_graph.py

```diff
@@ -460,187 +430,231 @@ def _split_classes_by_kind(self, class_name_to_definition):
+    elif type_id in COLLECTION_PROPERTY_TYPES:
+        if linked_class is not None and linked_type is not None:
+            raise AssertionError(u'Property "{}" unexpectedly has both a linked class and '
+                                 u'a linked type: {}'.format(name,
+                                                             orientdb_property_definition))
+        elif linked_type is not None and linked_class is None:
+            # No linked class, must be a linked native OrientDB type.
+            validate_supported_property_type_id(name + ' inner type', linked_type)
+
+            qualifier = linked_type
+        elif linked_class is not None and linked_type is None:
+            # No linked type, must be a linkz-graph user-defined type.
+            if linked_class not in non_graph_class_names:
+                raise AssertionError(u'Property "{}" is an embedded collection but has '
+                                     u'neither a linked class nor a linked type: '
+                                     u'{}'.format(name, linked_class))
+            qualifier = linked_class
+        else:
+            raise AssertionError(u'Property "{}" is an embedded collection but has '
+                                 u'neither a linked class nor a linked type: '
+                                 u'{}'.format(name, orientdb_property_definition))
+
+    default_value = None
+    default_value_string = orientdb_property_definition.get('defaultValue', None)
+    if default_value_string is not None:
+        default_value = parse_default_property_value(name, type_id, default_value_string)
+
+    descriptor = PropertyDescriptor(type_id=type_id, qualifier=qualifier, default=default_value)
+    # Sanity-check the descriptor before returning it.
+    _validate_collections_have_default_values(class_name, name, descriptor)
+    return descriptor
+
+
+def _try_get_leaf_link(class_name, link_direction, links, abstract,
+                       subclass_sets):
+    """Try to get the leaf link of an edge's end.
+
+    Args:
+        class_name: string, the name of the edge class.
+        link_direction: string, either 'in' or 'out', describing the edge's end of interest.
+        links: dict, string -> [string], mapping a link direction to a list set of
+               classes that can appear on that end of the edge class. The link
+               direction can be one of 'in' or 'out'.
+        abstract: bool, describes whether the class is abstract.
+        subclass_sets: dict, string -> [string], mapping a class to its subclasses.
+
+    Returns:
+        optional string, describing the leaf link. We define the links of an edge's end to be the
+        set of classes that could appear in that end of the edge. We define the leaf link as the
+        link that is not inherited by any of the other links. Each edge's end can have at
+        most 1 leaf link. The ends of non-abstract edges must each have a leaf link.
```

Comment:
I still think this is not clear.
1. The name leaf is inappropriate because subclass relationships don't form a tree.
2. I can't see how "the link that is not inherited by any of the other links" is useful in constructing the schema.
3. Can you justify some facts? Why are we guaranteed to find at most one leaf link? Why do the ends of non-abstract edges have to have a leaf link?

The way I understanding this by reading the code is we want to derive the end type of an edge based on the end types of its subclasses. Can we have the documentation (and function name) reflect that, and then as an implementation detail, mention how it's constructed?